code/modules/reagents/chemistry/machinery/chem_separator.dm

```diff
@@ -0,0 +1,239 @@
+	..()
+	if(A == beaker)
+		beaker = null
+		update_appearance()
+
+/obj/structure/chem_separator/update_overlays()
+	. = ..()
+	set_light(burning ? 1 : 0)
+	// Burner overlay
+	if(burning)
+		. += mutable_appearance(icon, "[icon_state]_burn", alpha = alpha)
+		. += emissive_appearance(icon, "[icon_state]_burn", src, alpha = alpha)
+	// Separator reagents overlay
+	if(reagents.total_volume)
+		var/threshold = null
+		for(var/i in 1 to fill_icon_thresholds.len)
+			if(ROUND_UP(100 * reagents.total_volume / reagents.maximum_volume) >= fill_icon_thresholds[i])
+				threshold = i
+		if(threshold)
+			var/fill_name = "[fill_icon_state]_m_[fill_icon_thresholds[threshold]]"
+			var/mutable_appearance/filling = mutable_appearance(fill_icon, fill_name)
+			filling.color = mix_color_from_reagents(reagents.reagent_list)
+			. += filling
+	// Beaker overlay
+	if(beaker)
+		. += "[icon_state]_beaker"
+		// Beaker reagents overlay
+		if(beaker.reagents.total_volume)
+			var/threshold = null
+			for(var/i in 1 to fill_icon_thresholds.len)
+				if(ROUND_UP(100 * beaker.reagents.total_volume / beaker.reagents.maximum_volume) >= fill_icon_thresholds[i])
+					threshold = i
+			if(threshold)
+				var/fill_name = "[fill_icon_state]_b_[fill_icon_thresholds[threshold]]"
+				var/mutable_appearance/filling = mutable_appearance(fill_icon, fill_name)
+				filling.color = mix_color_from_reagents(beaker.reagents.reagent_list)
+				. += filling
+	// Thermometer overlay
+	var/threshold = null
+	for(var/i in 1 to temperature_icon_thresholds.len)
+		if(ROUND_UP(reagents.chem_temp - T0C) >= temperature_icon_thresholds[i])
+			threshold = i
+	if(threshold)
+		var/fill_name = "[icon_state]_temp_[temperature_icon_thresholds[threshold]]"
+		var/mutable_appearance/filling = mutable_appearance(icon_state, fill_name)
+		. += filling
+
+/obj/structure/chem_separator/proc/burn_attackby_check(obj/item/I, mob/living/user)
+	var/ignition_message = I.ignition_effect(src, user)
+	if(!ignition_message)
+		return
+	. = TRUE
+	user.visible_message(ignition_message)
+	fire_act(I.get_temperature())
+
+/obj/structure/chem_separator/attackby(obj/item/I, mob/user, params)
+	if(burn_attackby_check(I, user))
+		return
+	if(is_reagent_container(I) && !(I.item_flags & ABSTRACT) && I.is_open_container())
+		var/obj/item/reagent_containers/new_beaker = I
+		if(!user.transferItemToLoc(new_beaker, src))
+			return
+		replace_beaker(user, new_beaker)
+		to_chat(user, span_notice("You add [new_beaker] to [src]."))
+		update_appearance()
+		. = TRUE // no afterattack
+	else
+		return ..()
+
+/obj/structure/chem_separator/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
+	if(!user)
+		return FALSE
+	if(beaker)
+		if(burning)
+			stop()
+		if(!issilicon(user) && in_range(src, user))
+			user.put_in_hands(beaker)
+		else
+			beaker.forceMove(drop_location())
+		beaker = null
+	if(new_beaker)
+		beaker = new_beaker
+	update_appearance()
```

Comment:
These should all be `update_appearance(UPDATE_ICON)` so we don't have any load bearing name/desc/similar updates.